### PR TITLE
Support downgrade to specific version in yum module.

### PR DESCRIPTION
**Pull request by ademmers**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
yum module

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel dbb452100a)
```

##### SUMMARY

Add possibility to downgrade a package via yum when a version lower than a current, possibly, installed version is provided.
This is a rather shameless copy of the code suggested by @willusher on the old core module code base.
(cfr: [ansible-modules-core#1419](https://github.com/ansible/ansible-modules-core/issues/1419) and [his commit](https://github.com/willusher/ansible-modules-core/commit/7fe933990cc9355104d5d56bdda54af56a511cc8) )

PRE:
```
$ ansible <host>-sKo -m yum -a 'name=<package>-<lower version as installed on host> state=installed'
SUDO password:
<host>| SUCCESS => {"changed": false, "msg": "", "rc": 0, "results": ["<package>-<lower version as installed on host>: Nothing to do"]}
```

POST:
```
$ ansible <host>-sKo -m yum -a 'name=<package>-<lower version as installed on host> state=installed'
SUDO password:
<host>| SUCCESS => {"changed": true, "msg": "", "rc": 0, "results": ["<very large yum output, depending on the actual package hopefully ending with Completed!>"]}
```

### CODE REVIEW

**Comment by vmindru** [comment]
I think this is a bad approach, it would be more consistent to have an option downgrade that you can specify and trigger the downgrade mechanism. Trying to analyse package versions and "guess" if the desired result is downgrade is a bit too brave imho.

**Comment by ademmers** [comment]
@vmindru : You can absolutly be right! Kind of needed a solution to downgrade on short notice and came across this one on the old core modules repository, with no exact reasoning why it was never considered for merge.
I will open an issue next stating the need for yum downgradebility and reference this one as a possible solution. 
Later (when finding some spare time), I'll have a look at the possibility of having a downgrade option (or someone else).

On the other side, I kind of like the idea of just specifying which package and version I would like to have actively installed, and the module takes care of it (aka without the need of any, possible extra option)

**Comment by vmindru** [comment]
the idea is good I am just afraid you are forcing one into exact versioning format which is not necessarily true to happen. E.G. yum package manager is used by multiple distros with the different point of view on how to manage version labeling.

**Comment by dagwieers** [comment]
@vmindru The RPM version label is pretty hardwired for all RPM-based distributions, so I don't think this is really a problem. My concern (with this implementation) is related to not using yum for making decisions, and the fact that the flow can do two actions (downgrade and upgrade) which basically causes the results `res` to get lost for the first action (downgrade).

So the right approach IMO is to instruct yum with the right information and make it do downgrades as part of the normal upgrade action (that sounds wrong, but downgrade simply is a special upgrade case).

**Comment by vmindru** [comment]
@dagwieers I am a bit confused, is the takeaway here: we should have a separate downgrade option as part of this module or not?

**Comment by dagwieers** [comment]
@vmindru Not.

**Comment by dagwieers** [comment]
@vmindru The aim for Ansible is to describe your desired state, not define specific actions. So in this case your desired state is "I want package A with version X.Y.Z, and package B". If A has Y.Z installed, and X.Y.Z is lower, Ansible should simply perform a downgrade to get to the desired state. Obviously that can be dangerous, and maybe the module should give a warning just in case (warnings can be suppressed).

Maybe even an option "allow_downgrade" which defaults to "no", to ensure that the user knows he wants to do something atypical.

**Comment by vmindru** [comment]
+1 on the allow_downgrade option.

**Comment by ademmers** [comment]
+1 on allow_downgrade
(maybe strange to do on own PR :), but hey, I do agree )

@dagwieers Can you provide me some more info on "yum making the descision" ?

**Comment by dagwieers** [comment]
Ok, so Yum already can do downgrades (using the downgrade option). And this also enables the user to downgrade to specific versions. What I think the module should do, is do installations and downgrades in the same transaction.

So if you provide it a list of packages that you expect to be installed (desired state) some of these may have exact version numbers specified and yum should be able to handle that, preferably internally in a single transaction. It is possible, because if you have package-conflicts you can tell Yum to resolve them for you (as good as possible) and it wil downgrade package at will too.

The user may not know if adding specific versions to their action will cause a downgrade or not, and that is the reason to have an option to allow downgrades, rather than make it a specific action (which is harder to handle correctly/idempotently in relation with other options, like a package without a specific version).

I didn't say that the implementation was going to be easy :-) You'll probably need to know a lot about the internals of yum, and rpmlib too.

**Comment by ademmers** [comment]
I see ... then that will might have to wait a bit :-)

I however already did just add the allow_downgrade option, defaults to False, so nothing will break for anyone not stating their downgrade intent.

**Comment by maratsh** [comment]
+1 for allow_donwgrade

**Comment by ReSearchITEng** [comment]
IMHO the module should behave similar to yum, and should not inhibit functionalities, especially important ones like downgrade.

**Comment by dagwieers** [comment]
needs_contributor

**Comment by mmercer-apixio** [comment]
In this particular case, I think that having the feature itself is more important than the manner in which it is achieved (so long as it does not *break* anything that currently exists in the module).  Given the age of the original submission for this feature from 2+ years ago, it would stand to reason that having the feature would be a beneficial gain, and the code can then be optimized and refactored as needed and time allows.  @abadger Thoughts on this one?  Would love to see you guys discuss this one as a consideration for merge into core

**Comment by abadger** [comment]
Adding a new parameter into a module with a stableinterface is something to be evaluated carefully.  We can't remove the option if it turns out to have been the wrong approach.  Maybe putting it on the meeting agenda to discuss what the parameter should look like is the right thing to do: https://github.com/ansible/community/issues/162

**Comment by mgcleveland** [comment]
+1 for allow_donwgrade

**Comment by mmercer-apixio** [comment]
Not liking that this keeps on getting 'buried' so to speak.  It adds support for a much needed, long time necessary feature, and every time someone from the community introduces a similar patch, it gets buried for years to the point of the patch being useless without complete rewrite.  Need to take actual action to prevent that from happening yet again for something that would be a huge gain.  I'd rather have the feature and have it not be 'perfectly ironed out' than not have it at all.

**Comment by Jmainguy** [comment]
yum downgrades are dangerous and wonky, however adding allow_downgrade would removed that concern for me, as someone would have to do this intentionally. I am ok with this getting merged in with allow_downgrade once this is rebased and tested.

**Comment by dagwieers** [comment]
I personally don't feel very comfortable with the "if it works we should add it" because if none of your are willing to do the work, none of you are actually going to support it when people are complaining that it doesn't work. I'd much rather have no implementation, than a lousy working implementation that gives us lots of complaining users and nobody that fixing it. (And don't tell me this is speculation, we have the unarchive module to remind me of this every day)

So if you feel powerless because there is functionality you really need, but someone is not giving it to you (like, *this* instance), go teach yourself a bit of python and make your hands dirty. It is not that difficult once you dive into it. That is how Open Source works, and why it's so powerful. You *can* make a difference.

PS ~The patch is not that hard to rebase~ (it apparently was) if anyone would care to do it, but IMO that still does not warrant we should merge it.

**Comment by Jmainguy** [comment]
So is that a hard no from you for this feature unless a yum module maintainer volunteers? Or is there any other way of this getting merged that would work for you? I agree I am not going to personally support this if it gets merged in, so take my +1 with that large grain of salt.

**Comment by dagwieers** [comment]
@Jmainguy I was asked by @ademmers (the original author) to review this. Apart from a voice I have no other authority, and no commit rights.

But yes, the lack of an active module maintainer plays a role in this functionality not being merged.

**Comment by Jmainguy** [comment]
Oh for sure, neither do I =), but your opinion means a lot to me. Thanks for looking at this.

**Comment by mmercer-apixio** [comment]
As with several others here, I would like to see the feature, but that is not the part I am trying to push in this.  I am trying to push forward the agenda of *discussing* what needs to be done, so that once we have established the 'what', we can tackle the 'how'.  Without knowing and understanding the what, we can't begin to solve the how.  I have already rebased the existing code, but I do agree with dag that simply merging it because it is done could likely be the wrong thing to do in this case.  Unfortunately, we also don't have a path forward, which makes changing it in an appropriate manner equally difficult.  That is the issue I would like to see discussed and resolved.  I will hold onto the rebase for now, so that we can try and establish a bit more communication and engagement on this.  I do agree that the logic side needs to happen internally in yum, that's why its there, in order to handle and solve deps and deptrees.

Going to take some of the feedback from dag, and try to put together an example of how it would look in a task, once we can define that, then we may be able to define the behavior as well.  Will dig through some of the existing parts of ansible and see what and how we handle version specification in things (for instance, pip package versions, rpm package versions, etc).  Maybe from there it will give us a baseline on how we can move forward and integrate it properly.

**Comment by mmercer-apixio** [comment]
dag took a look at some of the code and made an observation that I hadn't yet -- we're invoking yum, not the yum shell, so it doesn't natively support 'downgrade'.  It would support loading a transaction set, and we might be able to hook into that.  Will look into this and see if there is a way we can setup the transaction state and then load that.

**Comment by jctanner** [comment]
@mkrizek can you review this?

File: lib/ansible/modules/packaging/os/yum.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/yum.py]: After much discussion, we've decided that the overall design here (with a transaction for install and a transaction for downgrade) is okay.  But the documentation is very important.  So this entry needs to call out that specifying ```allow_downgrade=True``` means that the task is no longer idempotent.  The task could end up with a set of packages that does not match the complete list of specified packages to install (because dependencies between the downgraded package and others can cause changes to the packages which were in the earlier transaction). So please call that out here.  There name: field should also get some more documentation but I'll call it out there.

You can also update the version_added: to "2.4" here.

File: lib/ansible/modules/packaging/os/yum.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/yum.py]: In light of needing to set allow_downgrade, maybe something like

```
      - Package name, or package specifier with version, like C(name-1.0).
	If a previous version is specified, the task also needs to turn
	C(allow_downgrade) on. See the C(allow_downgrade) documentation for
	caveats with downgrading packages.  A previous version may be
	specified C(name-0.9) to downgrade a package.  When using
	state=latest, this can be '*' which means run C(yum -y update).  You
	can also pass a url or a local path to a rpm file (using
	state=present).  To operate on several packages this can accept
	a comma separated list of packages or (as of 2.0) a list of packages.
```

**Comment by abadger** [comment]
Okay, we've debated and discussed the overall design here and are okay with it as long as it has proper documentation (commented inline about what we want).  mkriezek will do a more in-depth review.

**Comment by abadger** [comment]
Note that this module is going into feature freeze for the 2.4.x release next week.  So, as dag said earlier, a contributor actively working on this or not will be important in determining whether this makes it in for 2.4.x or has to wait for 2.5.x.

**Comment by dagwieers** [comment]
@ademmers You can rescue the situation, and save the girl !

**Comment by mkrizek** [comment]
@ademmers Can you please rebase again? With your current commit you threw away this change https://github.com/ansible/ansible/commit/1c4e491eac3228bad51d3710bbed4d1ebdc15764. Thanks.

File: lib/ansible/modules/packaging/os/yum.py
Comment (mkrizek) [inline lib/ansible/modules/packaging/os/yum.py]: Ha, there seemed to be a bug in the previous code, we might want to do
```
for spec in items
    if spec.startswith('@'):
       ...
```

File: lib/ansible/modules/packaging/os/yum.py
Comment (mkrizek) [inline lib/ansible/modules/packaging/os/yum.py]: Shouldn't  `pkgs.append(pkg)` be removed? As you do that again in the else clause below?

File: lib/ansible/modules/packaging/os/yum.py
Comment (mkrizek) [inline lib/ansible/modules/packaging/os/yum.py:840]: So if `downgrade_pkgs` is not empty and `pkgs` is not empty, you rewrite some of the `res` content here via `exec_install` (like `changed`), no?

**Comment by ademmers** [comment]
@mkrizek I Just fixed the rebase for this code, but for obvious reasons the build fails: the spec variable is not defined in exec_install function (where that piece of code moved to).

Somehow this was already making use of some python specifics around variable scope (cfr [small example](https://gist.github.com/ademmers/67a7200e25fa9248164f29d905de7dcd)

Please suggest a potential fix, I'm not seeing it right now.

**Review by mkrizek — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/packaging/os/yum.py
Comment (mkrizek) [inline lib/ansible/modules/packaging/os/yum.py]: Oh, that doesn't sound right, let me come up with a fix.

File: lib/ansible/modules/packaging/os/yum.py
Comment (ademmers) [reply lib/ansible/modules/packaging/os/yum.py]: true, also result of complex rebase ...

File: lib/ansible/modules/packaging/os/yum.py
Comment (ademmers) [reply lib/ansible/modules/packaging/os/yum.py:840]: yes, several res fields might indeed be rewritten

File: lib/ansible/modules/packaging/os/yum.py
Comment (mkrizek) [reply lib/ansible/modules/packaging/os/yum.py]: No worries :)

File: lib/ansible/modules/packaging/os/yum.py
Comment (mkrizek) [inline lib/ansible/modules/packaging/os/yum.py:805]: How about doing all of this below only `if allow_downgrade:`?

File: lib/ansible/modules/packaging/os/yum.py
Comment (abadger) [inline lib/ansible/modules/packaging/os/yum.py]: Oops, the following sentence is now redundant and can be removed:
> A previous version may be specified C(name-0.9) to downgrade a package.

File: lib/ansible/modules/packaging/os/yum.py
Comment (ademmers) [reply lib/ansible/modules/packaging/os/yum.py:805]: You are right, clearly a simple and good optimization.

File: lib/ansible/modules/packaging/os/yum.py
Comment (ademmers) [reply lib/ansible/modules/packaging/os/yum.py]: will remove the redundancy

**Review by mkrizek — APPROVED** [review]
At this point, the code looks good to me and works in my testing. There are some potential issues with running `downgrade` and `install` in separate transactions as are with the downgrade itself but users are warned in the module docs. I am not against merging it into 2.4 since the feature is not enabled by default and users should know what they are doing when they enable it. If we want the feature more tested than postponing it until 2.5 would be wise.

**Comment by abadger** [comment]
Merging to devel... bcoca and I felt that the various incarnations of the PR have had a lot of testing, but were waiting on us to review and approve.  jctanner was a little more on the conservative side since yum is a major package manager.  I've merged it to devel so that people can start using it.  If no regressions are found in the beta and rc period it will be in 2.4.0.  If regressions (or big bugs in the new feature) are found we can always revert for 2.4 and fix the code for 2.5 instead.

Thanks @ademmers and all the rest who have contributed code, testing, and reviews to this code over the course of many long releases!

**Pull request merged**
